### PR TITLE
[SPARK-39167][SQL] Throw an exception w/ an error class for multiple rows from a subquery used as an expression

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
In the PR, I propose to use the MULTI_VALUE_SUBQUERY_ERROR error classes for multiple rows from a subquery used as an expression.

### Why are the changes needed?
Porting the executing errors for multiple rows from a subquery used as an expression to the new error framework should improve user experience with Spark SQL.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added new test suite

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala [hunk 1]
```diff
@@ -1971,4 +1971,10 @@ object QueryExecutionErrors extends QueryErrorsBase {
         s"add ${toSQLValue(amount, IntegerType)} $unit to " +
         s"${toSQLValue(DateTimeUtils.microsToInstant(micros), TimestampType)}"))
   }
+
+  def multipleRowSubqueryError(plan: String): Throwable = {
+    new SparkIllegalStateException(
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryExecutionErrors.scala:1976]: Please, use SparkException or SparkRuntimeException. I removed the the exception by https://github.com/apache/spark/pull/36550.

BTW, we shouldn't throw `IllegalStateException` if the exception can be trigger by user code in regular cases.

**Review by MaxGekk — CHANGES_REQUESTED** [review]

**Review by MaxGekk — APPROVED** [review]
LGTM, @panbingkun could you re-trigger GAs via an empty commit:
```
$ git commit --allow-empty -m "Trigger build"
```

**Comment by panbingkun** [comment]
> git commit --allow-empty -m "Trigger build"

Done

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master.
Thank you, @panbingkun.

**Pull request closed without merging**